File: ajuste-curvas-minimos-quadrados/minimos_quadrados.py
# =========================================================================
# AJUSTE DE CURVAS: MÉTODO DOS MÍNIMOS QUADRADOS (PYTHON)
# Modelo Exponencial - Consumo de Energia Elétrica
# =========================================================================


import numpy as np
import matplotlib.pyplot as plt


x = np.arange(1, 13).reshape(-1, 1)
y = np.array([140, 220, 150, 170, 190, 180, 170, 160, 180, 200, 170, 199]).reshape(-1, 1)

phi1 = np.exp(-x)
phi2 = np.ones_like(x)
phi3 = np.exp(x)

P = np.hstack((phi1, phi2, phi3))

coeficientes, residuos, rank, s = np.linalg.lstsq(P, y, rcond=None)

c1, c2, c3 = coeficientes.flatten()

y_predito = P @ coeficientes
norma_residuo = np.linalg.norm(y_predito - y)**2
print(f"Resíduo Quadrático (Modelo Exponencial): {norma_residuo:.1f}")

x_continuo = np.linspace(0, 12, 100).reshape(-1, 1)
y_continuo = c1 * np.exp(-x_continuo) + c2 * np.ones_like(x_continuo) + c3 * np.exp(x_continuo)

plt.figure(figsize=(8, 5))
plt.scatter(x, y, color='blue', marker='x', label='Dados Reais', s=80)
plt.plot(x_continuo, y_continuo, color='magenta', label='Ajuste Exponencial')
plt.title('Ajuste de Curvas - Consumo KWh (Mínimos Quadrados)')
plt.xlabel('Mês')
plt.ylabel('Consumo (KWh)')
plt.legend()
plt.grid(True, linestyle='--', alpha=0.6)
plt.show()

File: ajuste-curvas-minimos-quadrados/minimos_quadrados.m
% =========================================================================
% AJUSTE DE CURVAS: MÉTODO DOS MÍNIMOS QUADRADOS (OCTAVE)
% Modelo Exponencial - Consumo de Energia Elétrica
% =========================================================================


x = [1; 2; 3; 4; 5; 6; 7; 8; 9; 10; 11; 12];
y = [140; 220; 150; 170; 190; 180; 170; 160; 180; 200; 170; 199];

phie1 = @(x) exp(-x);
phie2 = @(x) ones(length(x), 1);
phie3 = @(x) exp(x);

Pe = [phie1(x) phie2(x) phie3(x)];

% Ae * ce = be => (Pe' * Pe) * ce = Pe' * y
Ae = Pe' * Pe;
be = Pe' * y;
ce = Ae \ be;

phie_fit = @(xx) ce(1)*phie1(xx) + ce(2)*phie2(xx) + ce(3)*phie3(xx);

norma_exponencial = norm(phie_fit(x) - y)^2;
fprintf('Residuo Quadratico (Modelo Exponencial): %.1f\n', norma_exponencial);

xx = linspace(0, 12, 60)';
plot(x, y, 'x', xx, phie_fit(xx), 'm');
title('Ajuste Exponencial - Consumo de Energia');
xlabel('Mês');
ylabel('Consumo (KWh)');
legend('Dados Reais', 'Ajuste Exponencial');